File: days/day_23/main.py
import time
import os

class Cup:
	def __init__(self, num):
		self.num = num
		self.right = None

	def __str__(self):
		return str(self.num)

	def __repr__(self):
		return str(self.num)


def create_cups(nums):
	n = len(nums)
	
	cups = { nums[i]: Cup(nums[i]) for i in range(n)}

	for i in range(0,len(nums)):
		cups[nums[i]].right = cups[nums[(i+1)%n]]

	return cups


def update(current, cups, n):
	picked_up = []

	j = current.right
	for i in range(1,4):
		picked_up.append(j.num)
		j = j.right

	current.right = j

	dest = current.num - 1 if current.num > 1 else n

	while dest in picked_up:
		dest = dest - 1 if dest > 1 else n

	tmp = cups[dest].right
	cups[dest].right = cups[picked_up[0]]
	cups[picked_up[2]].right = tmp

	return (cups[current.right.num], cups)

def part_one(nums):
	cups = create_cups(nums)
	n = len(cups)
	current = cups[1]

	for i in range(100):
		current, cups = update(current, cups, n)

	j = cups[1].right
	s = ''
	
	for i in range(1, len(cups)):
		s += str(j.num)
		j = j.right
	
	return s


def part_two(nums):
	nums.extend([*range(10, 10**6 + 1)])
	cups = create_cups(nums)
	current = cups[nums[0]]
	n = len(cups)

	for i in range(10**7):
		current, cups = update(current, cups, n)

	total = 1
	j = cups[1].right
	for i in range(1, 3):
		total *= j.num
		j = j.right
	
	return total


def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		nums = [int(x) for x in data.readline()]

		part_one_ans = part_one(nums)
		part_two_ans = part_two(nums)

		print('day 23  ({:,.3f}s)'.format(time.time()-start_time))
		print('  part 1: {}'.format(part_one_ans))
		print('  part 2: {}'.format(part_two_ans))


if __name__ == "__main__":
		main()

File: days/day_23/main.js
const fs = require('fs');
const readline = require('readline');

async function loadInput() {
  const rl = readline.createInterface({
    input: fs.createReadStream(__dirname + '/input.txt')
  });

  const lines = [];
  for await (const line of rl) {
    lines.push(line);
  }
  return lines
}

class Cup {
	constructor(num) {
		this.num = num
		this.right = null
	}

	toString() {
		return String(this.num)
	}
}

function createCups(nums) {
	const n = nums.length

	const cups = {}
	for (let i = 0; i < n; i++) {
		cups[nums[i]] = new Cup(nums[i])
	}
	for (let i = 0; i < n; i++) {
		cups[nums[i]].right = cups[nums[(i+1)%n]]
	}

	return cups
}

function update(current, cups, n) {
	const picked_up = []

	let j = current.right
	for (let i = 0; i < 3; i++) {
		picked_up.push(j.num)
		j = j.right
	}

	current.right = j

	let dest = current.num > 1 ? current.num - 1 : n

	while (picked_up.includes(dest)) {
		dest = (dest > 1 ? dest - 1 : n)
	}

	const tmp = cups[dest].right
	cups[dest].right = cups[picked_up[0]]
	cups[picked_up[2]].right = tmp

	return [cups[current.right.num], cups]
}


function partOne(nums) {
	const n = nums.length
	let cups = createCups(nums)
	let current = cups[nums[0]]

	for (let i = 0; i < 100; i++) {
		[current, cups] = update(current, cups, n)
	}

	let j = cups[1].right
	let s = ''
	
	for (let i = 1; i < Object.keys(cups).length; i++) {
		s += String(j.num)
		j = j.right
	}
	
	return s
}

function partTwo(nums) {
	for (let i = 10; i < (10**6+1); i++) {
		nums.push(i)
	}
	const n = nums.length
	const start = nums[0]
	let cups = createCups(nums)
	let current = cups[start]

	for (let i = 0; i < 10**7; i++) {
		[current, cups] = update(current, cups, n)
	}

	let total = 1
	let j = cups[start].right
	for (let i = 0; i < 2; i++) {
		total *= j.num
		j = j.right
	}
	
	return total
}

async function main() {
	const start_time = process.hrtime() 
  
  const lines = await loadInput();
  const nums = lines[0].split('').map(x => Number(x))

  const part_one_ans = partOne(nums)
  const part_two_ans = partTwo(nums)
  
  const diff = process.hrtime(start_time)
  const total_time = (diff[0] + diff[1]/1e9).toFixed(3)
  
  console.log(`Day  1 (${total_time}s)`)
  console.log(`  Part 1: ${part_one_ans}`);
  console.log(`  Part 2: ${part_two_ans}`);
}

if (require.main === module) {
  main();
}